feat: remove native chat usage limits in favor of AI Gateway budgets (#27329)
coder/coder: docs/ai-coder/agents/platform-controls/usage-insights.md

# Spend Management

Coder provides admin-only controls for monitoring and controlling agent
spend: AI Gateway budgets and cost tracking.

## Budgets

Coder Agents spend is controlled by AI Gateway budgets, which cap all AI
Gateway usage (including Coder Agents chats) per user over the configured
budget period.

- **Group budgets**: set a budget for a group from the group's settings
  page. The deployment budget policy resolves which group budget applies
  when a user belongs to multiple budgeted groups.
- **Per-user overrides**: set a custom budget for an individual user,

>   attributed to one of their groups. Takes priority over group budgets.

Author: Verified and documented in `d50c74cfee9`. This was the most useful finding in the review.

Confirmed both halves of the gate:

- `enterprise/coderd/coderd.go:640` wraps the `/ai/budget` routes in `httpmw.RequireExperiment(api.AGPL.Experiments, codersdk.ExperimentAIGatewayCostControl)`, so the endpoints answer "functionality requires enabling the ... experiment".
- `GroupMembersPage.tsx:86-92` gates the budget UI on the same experiment, with `TODO(AIGOV-443): drop the experiment gate once cost control is stable`.
- `ExperimentsSafe` contains only `ExperimentMinimumImplicitMember`, so `ai-gateway-cost-control` is excluded even from `--experimental=*`.

The Budgets section now carries an IMPORTANT note stating that budgets require the experiment, and that without it the controls are hidden and the endpoints refuse, so no budget can be configured.

The release-sequencing question this raises is real and above my pay grade: merging this PR removes the native limits while the replacement is still experiment-gated. I have escalated that to Mike rather than deciding it. Documenting the gate is the part I can do correctly here.

> Mux investigated and fixed this on Mike's behalf.

### Enforcement

- The AI Gateway checks the user's current spend before forwarding each
  request. When spend meets or exceeds the budget, the request is rejected
  and the chat shows a terminal error explaining that the budget was
  exceeded.
- Brief overage is possible when concurrent requests are in flight, because
  cost is recorded after the LLM responds.

### User-facing status

When a budget applies, users see their current AI spend, the budget, and
the period reset date in the usage indicator on the Agents page.

## Cost tracking

Navigate to **Agents** > **Settings** > **Manage Agents** > **Spend**.

> This view shows deployment-wide AI Gateway costs with per-user drill-down.

Author: You are right, and I was wrong to defer this twice. Fixed in `d50c74cfee9`.

This PR was rewriting the cost table to `Requests` / `Sessions` and "deployment-wide AI Gateway costs" while the page it documents still calls `paginatedChatCostUsers` / `chatCostSummary` (which aggregate `chat_messages`) and `UsageTab` still renders `Messages` / `Chats`. That inaccuracy was introduced by this commit, so "the next PR deletes it" was not an adequate answer.

The Cost tracking section is back to the chat-cost wording from `main`. The only edit it keeps is dropping the native usage-limit references that this PR removes (the limit-progress bullet, and the usage-limits clause in the title-generation note). #27330 then deletes the section outright, so the gateway wording never describes the native view at any commit.

This also shrinks the diff: the cost-table churn is gone entirely.

> Mux fixed this on Mike's behalf.

> It requires the AI Governance add-on.

Author: You are right, and I was wrong to defer this twice. Fixed in `d50c74cfee9`.

This PR was rewriting the cost table to `Requests` / `Sessions` and "deployment-wide AI Gateway costs" while the page it documents still calls `paginatedChatCostUsers` / `chatCostSummary` (which aggregate `chat_messages`) and `UsageTab` still renders `Messages` / `Chats`. That inaccuracy was introduced by this commit, so "the next PR deletes it" was not an adequate answer.

The Cost tracking section is back to the chat-cost wording from `main`. The only edit it keeps is dropping the native usage-limit references that this PR removes (the limit-progress bullet, and the usage-limits clause in the title-generation note). #27330 then deletes the section outright, so the gateway wording never describes the native view at any commit.

This also shrinks the diff: the cost-table churn is gone entirely.

> Mux fixed this on Mike's behalf.

### Top-level view

A per-user rollup table with the following columns:

| Column             | Description                            |
|--------------------|----------------------------------------|
| Total cost         | Aggregate dollar spend for the user    |
| Requests           | Number of intercepted LLM requests     |

> | Sessions           | Number of distinct AI Gateway sessions |

Author: Same stack-layering shape, and your reading of this head is correct.

At `999d24a147` the page still calls `paginatedChatCostUsers` and `UsageTab` still renders the native columns, so the gateway wording here does describe data the shipped page does not expose yet. #27330 deletes this entire "Cost tracking" section rather than repointing it, because no replacement dashboard survives the stack (the AI settings sidebar has no Spend entry after that commit).

So the transient mismatch exists only between these two commits, which land together. Verified at `62c153a5740`: the section, the columns table, and the per-user detail description are all gone.

> Mux reviewed and replied to this on Mike's behalf.

# Spend Management

Coder controls agent spend with AI Gateway budgets, and surfaces the

> resulting spend to both admins and users.

Reviewer: ```suggestion
Coder controls agent spend with AI Gateway budgets, and surfaces the resulting spend to both admins and users.
```

Author: Used the suggested opening sentence. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

## Budgets

Coder Agents spend is controlled by AI Gateway budgets, which cap all AI
Gateway usage (including Coder Agents chats) per user over the budget

> period.

Reviewer: ```suggestion
Coder Agents spend is controlled by AI Gateway budgets, which cap all AI Gateway usage (including Coder Agents chats) per user over the budget period.
```

Author: Used the suggested AI Gateway budget summary. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

- **Group budgets**: set a budget for a group from the group's settings
  page. The deployment budget policy resolves which group budget applies
  when a user belongs to multiple budgeted groups.
- **Per-user overrides**: set a custom budget for an individual user,

>   attributed to one of their groups. Takes priority over group budgets.

Reviewer: ```suggestion
- **Group budgets**: set a budget for a group from the group's settings page. 
  The deployment budget policy resolves which group budget applies when a user belongs to multiple budgeted groups, which defaults to the group with the highest budget.
- **Per-user overrides**: set a custom budget for an individual user, attributed to one of their groups.
  Per-user overrides take priority over group budgets.
```

Author: Applied the group-budget and per-user override wording. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

The deployment flags `--ai-budget-policy` and `--ai-budget-period` currently
support only `highest` and `month`. `highest` selects the group with the
largest spend limit, and `month` resets spend at the start of each UTC
calendar month. A user who belongs to no budgeted group falls back to the
Everyone group, which has no budget unless one is set for it. There is no
deployment-wide budget amount, and a configured spend limit cannot exceed
$1,000,000 per member per period.

> [!IMPORTANT]
> Budget controls in the Coder UI, the group budget endpoints
> (`/api/v2/groups/{group}/ai/budget`), and the AI spend status and reporting
> endpoints all require the AI Gateway entitlement. No experiment is needed.
>
> Native chat usage limits are removed from the application. Existing native
> limit values are not migrated to AI Gateway budgets and are no longer

> > enforced. Configure AI Gateway budgets separately.

Reviewer: ```suggestion
> Budget controls in the Coder UI, the group budget endpoints (`/api/v2/groups/{group}/ai/budget`), and the AI spend status and reporting endpoints all require the AI Gateway entitlement. 
> No experiment is needed.
>
> Native chat usage limits are removed from the application.
> Existing native limit values are not migrated to AI Gateway budgets and are no longer enforced.
> Configure AI Gateway budgets separately.
```

Also, I can't seem to find our API docs on AI budget. Do you know if they exist? Do I need to build a generator to surface them?

Author: Applied the callout wording. The generated Enterprise API reference already documents the group budget `GET`, `PUT`, and `DELETE` operations, so I also linked those sections from this page. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

### Enforcement

- The AI Gateway checks the user's current spend before forwarding each
  request. When spend meets or exceeds the budget, the request is rejected
  and the chat shows a terminal error explaining that the budget was
  exceeded.
- Brief overage is possible when concurrent requests are in flight, because

>   cost is recorded after the LLM responds.

Reviewer: ```suggestion
- The AI Gateway checks the user's current spend before forwarding each request.
  When spend meets or exceeds the budget, the request is rejected and the chat shows a terminal error explaining that the budget was exceeded.
- Brief overage is possible when concurrent requests are in flight, because cost is recorded after the LLM responds.
```

Author: Applied the enforcement wording. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

### User-facing status

The usage indicator on the Agents page and the summary in the user menu both
show the signed-in user's current AI spend, their budget, and the period
reset date. Both appear only when the deployment has the AI Gateway

> entitlement.

Reviewer: ```suggestion
The usage indicator on the Agents page and the summary in the user menu both show the signed-in user's current AI spend, their budget, and the period reset date.
Both appear only when the deployment has the AI Gateway entitlement.
```

Author: Applied the user-facing status wording. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

## Spend visibility

Spend is shown where it is actionable:

- **Agents page and user menu**: the signed-in user's spend against their
  budget, as described above.
- **Group settings**: each member's spend against the group's budget, for
  admins who can manage the group.
- **Chat summary panel**: the cost of one chat tree, on a chat's Summary tab.
  A subagent reports the total for its whole tree, including the chat that
  started it.
- **Agents** > **Settings** > **Manage Agents** > **Spend**: deployment-wide

>   chat cost per user, with per-user drill-down.

Reviewer: ```suggestion
- **Agents page and user menu**: the signed-in user's spend against their budget, as described previously.
- **Group settings**: each member's spend against the group's budget, for admins who can manage the group.
- **Chat summary panel**: the cost of one chat tree, on a chat's Summary tab.
  A subagent reports the total for its whole tree, including the chat that started it.
- **Agents** > **Settings** > **Manage Agents** > **Spend**: deployment-wide chat cost per user, with per-user drill-down.
```

Author: Applied the spend-surface wording for this stack layer. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

> [!NOTE]
> Per-chat cost comes from AI Gateway records, which are pruned according to
> `--ai-gateway-retention` (60 days by default). A chat whose gateway records

> > have been pruned reports no cost.

Reviewer: ```suggestion
> Per-chat cost comes from AI Gateway records, which are pruned according to `--ai-gateway-retention` (60 days by default).
> A chat for which gateway records have been pruned reports no cost.
```

I avoid using "whose" when talking about things that aren't people. Using "for which" is usually cleaner.

Author: Changed the retention sentence to use `for which`. Applied in `a676bbede97`.\n\n> Mux applied this feedback on Mike's behalf.

> (`/api/v2/groups/{group}/ai/budget`) require the AI Gateway entitlement but
> do not require the experiment. Gateway enforcement still applies to
> configured budgets when the experiment is disabled.
>
> Native chat usage limits are removed from the application. Existing native
> limit values are not migrated to AI Gateway budgets and are no longer
> enforced. Configure AI Gateway budgets separately.

### Enforcement

- The AI Gateway checks the user's current spend before forwarding each
  request. When spend meets or exceeds the budget, the request is rejected
  and the chat shows a terminal error explaining that the budget was
  exceeded.
- Brief overage is possible when concurrent requests are in flight, because
  cost is recorded after the LLM responds.

### User-facing status

The usage indicator on the Agents page and the summary in the user menu both
show the signed-in user's current AI spend, their budget, and the period
reset date. Both appear only when the deployment has the AI Gateway
entitlement and the `ai-gateway-cost-control` experiment is enabled.

## Spend visibility

There is no dedicated deployment-wide spend dashboard. Spend is shown where

> it is actionable:

Author: Good catch, and correct on the facts: at this commit `site/src/router.tsx` still routes `/ai/settings/spend` and `SpendPageView` still renders the deployment-wide usage-by-user table, so claiming no such dashboard exists was wrong here.

Fixed in 6b52f14803e by moving that claim to the PR that actually removes the page. This PR now documents the Spend page as a live surface (`**Agents** > **Settings** > **Manage Agents** > **Spend**: deployment-wide chat cost per user, with per-user drill-down`) alongside the new gateway budget surfaces, and #27330 replaces that bullet with "There is no dedicated deployment-wide spend dashboard" in the same commit that deletes the route and the page.

Verified per branch: this head has `router-spend=3` with no dashboard denial, and #27330 has `router-spend=0` with the denial restored.

> Mux is acting on Mike's behalf.
